### PR TITLE
Support torch function user objects

### PR DESCRIPTION
**Pull request by mlazos**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #111765

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

**Review by jansel — CHANGES_REQUESTED** [review]
Tests are failing

File: torch/_dynamo/variables/user_defined.py [hunk 1]
```diff
@@ -204,11 +205,12 @@ class UserDefinedObjectVariable(UserDefinedVariable):
     Mostly objects of defined type.  Catch-all for something where we only know the type.
     """
 
-    def __init__(self, value, value_type=None, **kwargs):
+    def __init__(self, value, value_type=None, torch_function_fn=None, **kwargs):
         super().__init__(**kwargs)
         self.value = value
         self.value_type = value_type or type(value)
         assert type(value) is self.value_type
+        self.torch_function_fn = torch_function_fn
```
Comment (ezyang) [inline torch/_dynamo/variables/user_defined.py:213]: @jansel @voznesenskym so there's this pattern of passing in "inner" variable trackers during construction of variable trackers and it feels like this is counter to the upcoming lazy VT regime. But I am not sure what the correct style here is. Should the constructor here recursively call builder? Should this be forced to be a non-optional argument?

Comment (mlazos) [reply torch/_dynamo/variables/user_defined.py:213]: I originally had a separate variable tracker inheriting from UDOV, but Ed suggested just adding functionality to UDOV because (I think) it's more representative of the python object model. Curious what you guys think

Comment (jansel) [reply torch/_dynamo/variables/user_defined.py:213]: Shouldn't this be a method on UserDefinedObjectVariable as opposed to a property?

We should only check (and guard on) torch_function if the object is used in a place where having a torch function matters.  Doing it outside the constructor seems odd.

File: torch/_dynamo/variables/user_defined.py [hunk 2]
```diff
@@ -224,6 +226,22 @@ def __str__(self):
...
+
+        return call_torch_function(
+            tx,
+            _get_subclass_type_var(tx, self),
+            self.torch_function_fn,
+            fn,
+            types,
+            args,
+            kwargs,
+        )
```
Comment (voznesenskym) [inline torch/_dynamo/variables/user_defined.py:243]: This is rather hard to follow, can you explain what this is doing?

Comment (mlazos) [reply torch/_dynamo/variables/user_defined.py:243]: `call_torch_function` is the main entry point to inlining a `__torch_function__` impl, each class that supports `__torch_function__` arranges the arguments according to the `__torch_function__` signature and calls that to inline through the torch function var.

**Review by jansel — APPROVED** [review]
